File: nekoshlink/shlink-security/src/main/java/org/nekosoft/shlink/sec/delegation/RunAsDelegationAspect.java
package org.nekosoft.shlink.sec.delegation;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.springframework.stereotype.Component;

import java.util.function.Function;

@Aspect
@Component
public class RunAsDelegationAspect {

    @Around("@annotation(runAs)")
    public Object delegate(ProceedingJoinPoint joinPoint, org.nekosoft.shlink.sec.delegation.annotation.RunAs runAs) throws Throwable {

        Function<String[], RunAs> executor;
        if (runAs.allowAnonymous()) {
            executor = RunAs::anonymousWithRoles;
        } else {
            executor = RunAs::userWithRoles;
        }

        Object result;
        try (RunAs ignored = executor.apply(runAs.roles())) {
            result = joinPoint.proceed();
        }
        return result;
    }

}

File: nekoshlink/shlink-security/src/main/kotlin/org/nekosoft/shlink/sec/delegation/RunAsDelegationAspect.kt
package org.nekosoft.shlink.sec.delegation

import org.nekosoft.shlink.sec.delegation.annotation.RunAs

import org.nekosoft.shlink.sec.delegation.RunAs as RunAsBlock

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.springframework.stereotype.Component;

@Aspect
@Component
class RunAsDelegationAspect {

    @Around("@annotation(runAs)")
    fun delegate(joinPoint: ProceedingJoinPoint, runAs: RunAs): Any? {

        val executor = if (runAs.allowAnonymous) {
            RunAsBlock.Companion::anonymousWithRoles
        } else {
            RunAsBlock.Companion::userWithRoles
        }

        val result = executor(runAs.roles). use {
            joinPoint.proceed()
        }

        return result
    }

}
